From b6006a32aa3fff287a6159caa27c7b0db2dde193 Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Tue, 23 Oct 2018 14:06:25 +0800
Subject: [PATCH 4902/5242] MLK-20033 ARM64: dts: imx8qm: support android auto
 plus m4 on xen

commit  c68bc8a90bde423a607bc720d63c6bc62c9a3ed0 from
https://source.codeaurora.org/external/imx/linux-imx.git

Map M41 reserved DDR memory for rpmsg
Add init-smmu-bypass, because m41 is started by SCU at early stage,
configure sid for a running master needs some specific handle. For
CM41, we configure S2CR with BYPASS in init stage.
Add mipi csi gpio
Add can_rpmsg node, this node is only used by android auto kernel.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-mek-dom0.dts     |   11 ++++++++++
 .../boot/dts/freescale/fsl-imx8qm-mek-domu.dts     |   22 ++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
index 33ac207..15f6bad 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
@@ -186,10 +186,13 @@
 				SC_P_QSPI1A_DATA0
 				/* isl29023 */
 				SC_P_USDHC2_WP
+				SC_P_MIPI_CSI0_GPIO0_00
+				SC_P_MIPI_CSI0_GPIO0_01
 
 				>;
 			gpios = <&gpio1 13 GPIO_ACTIVE_LOW>,
 				<&gpio1 27 GPIO_ACTIVE_LOW>,
+				<&gpio1 28 GPIO_ACTIVE_LOW>,
 				<&gpio4 6 GPIO_ACTIVE_LOW>,
 				<&gpio4 9 GPIO_ACTIVE_LOW>,
 				<&gpio4 11 GPIO_ACTIVE_HIGH>,
@@ -235,6 +238,8 @@
 		#stream-id-cells = <1>;
 		iommus = <&smmu>;
 		xen,passthrough;
+		/* This will be parased by xen smmu driver */
+		init-smmu-bypass;
 	};
 
 	irqsteer_cm41: irqsteer_cm41@0x51080000 {
@@ -277,6 +282,12 @@
 		reg = <0 0x94400000 0 0x1800000>;
 	};
 
+	/* This piece memory is used by M41, M40 is not covered */
+	m41_mem: m41_mem@0x94400000 {
+		xen,passthrough;
+		reg = <0 0x88800000 0 0x7800000>;
+	};
+
 	gpio4_dummy: gpio4_dummy@0{
 		/* Passthrough gpio4 interrupt to DomU */
 		interrupts = <GIC_SPI 140 IRQ_TYPE_LEVEL_HIGH>;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
index 9ee3cfc..f0c1d22 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
@@ -171,6 +171,24 @@
 		#address-cells = <2>;
 		#size-cells = <2>;
 
+		/* Android Auto Kernel will use this */
+		can_rpmsg: can_rpmsg {
+			compatible = "nxp,imx-can-rpmsg";
+			power-domains = <&pd_dc1>;
+			/*
+			 * fsl,resources is used to open the power of these resources.
+			 * m4 will power down these resources when M4 core give out the control of camera/display.
+			 * rpmsg_can driver will power on these reources once A core take over the control of camera/display.
+			 */
+			fsl,resources = <SC_R_LVDS_1 SC_R_LVDS_1_I2C_0
+					 SC_R_DC_1 SC_R_DC_1_PLL_1
+					 SC_R_CSI_0_I2C_0 SC_R_ISI_CH0
+					 SC_R_ISI_CH1 SC_R_ISI_CH2
+					 SC_R_ISI_CH3 SC_R_CSI_0>;
+			fsl,resources-num = <10>;
+			status = "disabled";
+		};
+
 		firmware {
 			android {
 				compatible = "android,firmware";
@@ -983,3 +1001,7 @@
 		};
 	};
 };
+
+&can_rpmsg {
+	status = "okay";
+};
-- 
1.7.9.5

